Mathlib/Topology/CWComplex/Classical/Subcomplex.lean

```diff
@@ -0,0 +1,180 @@
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Floris van Doorn, Hannah Scholz
+-/
+
+import Mathlib.Topology.CWComplex.Classical.Finite
+import Mathlib.Analysis.Normed.Module.RCLike.Real
+
+/-!
+# Subcomplexes
+
+In this file we discuss subcomplexes of CW complexes.
+The definintion of subcomplexes is in the file `Topology.CWComplex.Classical.Basic`.
+
+## Main results
+* `RelCWComplex.Subcomplex.instRelCWComplex`: a subcomplex of a (relative) CW complex is again a
+  (relative) CW complex.
+
+## References
+* [K. Jänich, *Topology*][Janich1984]
+-/
+
+noncomputable section
+
+open Metric Set
+
+namespace Topology
+
+variable {X : Type*} [t : TopologicalSpace X] {C D : Set X}
+
+/-- A subcomplex is the union of its closed cells and its base. -/
+lemma RelCWComplex.Subcomplex.union_closedCell [T2Space X] [RelCWComplex C D] (E : Subcomplex C) :
+    D ∪ ⋃ (n : ℕ) (j : E.I n), closedCell (C := C) n j = E := by
+  apply subset_antisymm
+  · apply union_subset
+    · rw [← Subcomplex.union (C := C) (D := D) (E := E)]
+      exact subset_union_left
+    · apply iUnion_subset fun n ↦ iUnion_subset fun i ↦ ?_
+      rw [← closure_openCell_eq_closedCell, E.closed.closure_subset_iff, ← E.union]
+      apply subset_union_of_subset_right
+      exact subset_iUnion_of_subset n
+        (subset_iUnion (fun (i : ↑(E.I n)) ↦ openCell (C := C) (D := D) n ↑i) i)
+  · simp_rw [← E.union]
+    apply union_subset_union_right
+    apply iUnion_mono fun n ↦ iUnion_mono fun (i : ↑(E.I n)) ↦ ?_
+    exact openCell_subset_closedCell (C := C) n i
+
+/-- A subcomplex is the union of its closed cells. -/
+lemma CWComplex.Subcomplex.union_closedCell [T2Space X] [CWComplex C] (E : Subcomplex C) :
+    ⋃ (n : ℕ) (j : E.I n), closedCell (C := C) n j = E :=
+  (empty_union _ ).symm.trans (RelCWComplex.Subcomplex.union_closedCell E)
+
+lemma RelCWComplex.Subcomplex.disjoint_openCell_subcomplex_of_not_mem [RelCWComplex C D]
+    (E : Subcomplex C) {n : ℕ} {i : cell C n} (h : i ∉ E.I n) : Disjoint (openCell n i) E := by
+  simp_rw [← union, disjoint_union_right, disjoint_iUnion_right]
+  exact ⟨disjointBase n i , fun _ _ ↦ disjoint_openCell_of_ne (by aesop)⟩
+
+open Classical in
+/-- A subcomplex is again a CW complex. -/
+@[simps]
+instance RelCWComplex.Subcomplex.instRelCWComplex [T2Space X] [RelCWComplex C D]
+    (E : Subcomplex C) : RelCWComplex E D where
+  cell n := E.I n
+  map n i := map (C := C) n i
+  source_eq n i := source_eq (C := C) n i
+  continuousOn n i := continuousOn (C := C) n i
+  continuousOn_symm n i := continuousOn_symm (C := C) n i
+  pairwiseDisjoint' := by
+    simp only [PairwiseDisjoint, Set.Pairwise, mem_univ, ne_eq, Function.onFun,
+      disjoint_iff_inter_eq_empty, true_implies, Sigma.forall, Subtype.forall]
+    intro n j _ m i _ eq
+    refine (disjoint_openCell_of_ne ?_).inter_eq
+    simp_all only [Sigma.mk.inj_iff, not_and, ne_eq]
+    intro a
+    subst a
+    simp_all
```

Comment:
```suggestion
    intro ⟨n, i⟩ _ ⟨m, j⟩ _ hne
    refine @pairwiseDisjoint' _ _ C D _ ⟨n, i⟩ trivial ⟨m, j⟩ trivial ?_
    exact Function.injective_id.sigma_map (fun _ ↦ Subtype.val_injective) |>.ne hne
```
Although I agree that `disjoint_openCell_of_ne` is generally a better interface, in this case I think it's just more natural to appeal to `pairwiseDisjoint'` directly.